scratchpad/tn_api/tn.py

```diff
@@ -0,0 +1,202 @@
+                if idx in tn._indices:
+                    tn._indices[idx].append(port)
+                else:
+                    tn._indices[idx] = [port]
+
+        for idx, slice_val in slice_dict.items():
+            # don't care about indices that are not in TN
+            if idx not in tn._indices:
+                continue
+            
+            edge = tn._edges.pop(idx)
+            # get all tensors indexed by this edge
+            tensors_to_slice = set(port.tensor_ref for port in edge)
+            # store slice index and value for each tensor
+            local_slices_dict = {}
+            for current_tensor_ref in tensors_to_slice:
+                slice_dict = {}
+                # get all ports for the current tensor
+                current_tensor_ref_ports = [port for port in edge where port.tensor_ref == current_tensor_ref]
+                for current_port in current_tensor_ref_ports:
+                    slice_dict[current_port.ix] = slice_val
+                # store the slice params for this tensor in the local dict
+                local_slices_dict[current_tensor_ref] = slice_dict
+
+            # now use the local slice dict to slice for each tensor
+            for current_tensor_ref, slice_dict in local_slices_dict.items():
+                slice_bounds = []
+                current_tensor = tn._tensors[current_tensor_ref]
+                for idx in range(current_tensor.ndim):
+                    try:
+                        slice_bounds.append(slice_dict[idx])
+                    except KeyError:
+                        slice_bounds.append(slice(None))
+                tn._tensors[current_tensor_ref] = tn._tensors[current_tensor_ref][tuple(slice_bounds)]
+
+        return tn
+
+    def copy(self):
+        new = TensorNetwork()
+        new._tensors = self._tensors
+        new._edges = self._edges
+        new.shape = self.shape
+        new.indices = self.indices
+        return new
+
+    def add(self, other: "TensorNetwork | np.ndarray"):
+        if not isinstance(other, TensorNetwork):
+            self._tensors.append(other)
+            self.shape = self.shape + other.shape
+        else:
+            m = len(self._tensors)
+            n = len(self.shape)
+            # -- other's edges tensors will refer to shifted tensor location
+            enew = []
+            for e in other._edges:
+                e_ = []
+                for p in e:
+                    if p.tensor_ref == -1:
+                        e_.append(Port(tensor_ref=-1, ix=p.ix+n))
+                    else:
+                        e_.append(Port(tensor_ref=p.tensor_ref+m, ix=p.ix))
+                enew.append(tuple(e_))
+
+            self._edges += tuple(enew)
+            self._tensors += other._tensors
+            self.shape += other.shape
+
+    # contract to produce a new tensor
+    def contract(self, contraction_info: ContractionInfo) -> np.ndarray:
+        raise NotImplementedError()
```

Comment:
contraction_info here is indices i'd want in the output
einsum

Reply:
test case is contracted A = contraction of sliceA[0] and sliceA[1]. this is true for any index so can select it at random